node/state.re

```diff
@@ -243,7 +200,36 @@ let load_snapshot =
       last_applied_block_timestamp: 0.0,
       last_seen_membership_change_timestamp: 0.0,
     };
-  let next_state_root = (state_root_hash, state_root);
-  let t = {...t, next_state_root, protocol, block_pool};
-  List.fold_left_ok(apply_block, t, all_blocks);
+  // We should only ever load snapshots of the future, never the past.
+  let.assert () = (
+    `Invalid_snapshot_height,
+    protocol.block_height > t.protocol.block_height,
+  );
+  let t = {...t, protocol, block_pool};
+  // TODO: it doesn't seem valid to add a snapshot here based on the information received
+  // from our (untrusted) peer; however, that's exactly what we're doing here. Supposing
+  // the peer sending the snapshot is dishonest, we will then start propogating a bad snapshot.
+  // In the future, we should only add snapshots once we're in sync.
+  List.fold_left_ok(
+    (prev_state, block) => {
+      let.ok state = apply_block(prev_state, block);
+      if (BLAKE2B.equal(
+            prev_state.protocol.state_root_hash,
+            state.protocol.state_root_hash,
+          )) {
+        Ok(state);
+      } else {
+        let (hash, data) = Protocol.hash(prev_state.protocol);
```

Comment:
This should definitely be in parallel. Otherwise in a tight situation the node will not be able to load the snapshot fast enough.

Reply:
I added a `TODO:`